### PR TITLE
Simplify hashToIndex method

### PR DESCRIPTION
**Pull request by erdinctaskin**
Eliminate redundant modular aritmetic for Integer.MIN_VALUE

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/util/HashUtil.java [hunk 1]
```diff
@@ -334,12 +334,10 @@ public static int hashToIndex(int hash, int length) {
...
 
         if (hash == Integer.MIN_VALUE) {
-            hash = 0;
-        } else {
-            hash = abs(hash);
+            return 0;
         }
 
-        return hash % length;
+        return abs(hash) % length;
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/util/HashUtil.java]: According to [Math#abs javaDoc](https://docs.oracle.com/javase/6/docs/api/java/lang/Math.html#abs(int))  `Integer.MIN_VALUE` is a special case and returns itself. Doesn't this change break the behaviour of the method?

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/util/HashUtil.java]: Oops I see now, you are right.

**Comment by pveentjer** [comment]
Always in favor of simplified code.

**Comment by pveentjer** [comment]
The following might be simpler

```
    if(hash<0){
            if(hash==Integer.MIN_VALUE){
                return 0;
            }else{
                return (-hash)%length;
            }
        }else{
            return hash % length;
        }
```

In the old code, for a positive hash, there are 2 if statement and have no been reduced to one.
In the old code, for negative hash, there are still 2 if statements.

In our case we frequently use a positive 'hash' e.g. partition id. So this would make these cases faster.

Let me run a simple JMH benchmark

**Comment by mmedenjak** [comment]
Hi @erdinctaskin ! Thank you for your commit.
Can you sign the contributor agreement and send it to us?
https://hazelcast.atlassian.net/wiki/spaces/COM/pages/6357071/Hazelcast+Contributor+Agreement

**Comment by pveentjer** [comment]
I added a quick JMH benchmark

```
package com;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Measurement;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OperationsPerInvocation;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.annotations.Warmup;
import org.openjdk.jmh.infra.Blackhole;

import java.util.concurrent.TimeUnit;

import static com.hazelcast.util.Preconditions.checkPositive;
import static com.hazelcast.util.QuickMath.modPowerOfTwo;
import static java.lang.Math.abs;


@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@State(Scope.Thread)
@Fork(value = 1, warmups = 1)
@Warmup(iterations = 2)
@Measurement(iterations = 5)
public class HashUtilBenchmark {

    @OperationsPerInvocation(100)
    @Benchmark
    public void hash_old_positive(Blackhole hole) {
        for (int k = 0; k < 100; k++) {
            hole.consume(hashToIndex_old(k, 271));
        }
    }

    @OperationsPerInvocation(100)
    @Benchmark
    public void hash_old_negative(Blackhole hole) {
        for (int k = 0; k < 100; k++) {
            hole.consume(hashToIndex_old(-k, 271));
        }
    }

    @OperationsPerInvocation(100)
    @Benchmark
    public void hash_rdinctaskin_positive(Blackhole hole) {
        for (int k = 0; k < 100; k++) {
            hole.consume(hashToIndex_rdinctaskin(k, 271));
        }
    }

    @OperationsPerInvocation(100)
    @Benchmark
    public void hash_rdinctaskin_negative(Blackhole hole) {
        for (int k = 0; k < 100; k++) {
            hole.consume(hashToIndex_rdinctaskin(-k, 271));
        }
    }

    @OperationsPerInvocation(100)
    @Benchmark
    public void hash_peter_positive(Blackhole hole) {
        for (int k = 0; k < 100; k++) {
            hole.consume(hashToIndex_peter(k, 271));
        }
    }

    @OperationsPerInvocation(100)
    @Benchmark
    public void hash_peter_negative(Blackhole hole) {
        for (int k = 0; k < 100; k++) {
            hole.consume(hashToIndex_peter(-k, 271));
        }
    }

    @OperationsPerInvocation(100)
    @Benchmark
    public void hash_peterOptimized_positive(Blackhole hole) {
        for (int k = 0; k < 100; k++) {
            hole.consume(hashToIndex_peter_optimized(k, 256));
        }
    }

    @OperationsPerInvocation(100)
    @Benchmark
    public void hash_peterOptimized_negative(Blackhole hole) {
        for (int k = 0; k < 100; k++) {
            hole.consume(hashToIndex_peter_optimized(-k, 256));
        }
    }

    private static int hashToIndex_old(int hash, int length) {
        checkPositive(length, "length must be larger than 0");

        if (hash == Integer.MIN_VALUE) {
            hash = 0;
        } else {
            hash = abs(hash);
        }

        return hash % length;
    }

    private static int hashToIndex_peter(int hash, int length) {
        checkPositive(length, "length must be larger than 0");

        if (hash < 0) {
            if (hash == Integer.MIN_VALUE) {
                return 0;
            } else {
                return (-hash) % length;
            }
        } else {
            return hash % length;
        }
    }

    private static int hashToIndex_rdinctaskin(int hash, int length) {
        checkPositive(length, "length must be larger than 0");


        if (hash == Integer.MIN_VALUE) {
            return 0;
        }

        return abs(hash) % length;
    }

    // length needs to be power of 2 to do fast mod.
    private static int hashToIndex_peter_optimized(int hash, int length) {
        if (hash < 0) {
            if (hash == Integer.MIN_VALUE) {
                return 0;
            } else {
                return modPowerOfTwo(-hash, length);
            }
        } else {
            return modPowerOfTwo(hash,length);
        }
    }
}
```

Results don't matter.

```
Benchmark                                       Mode  Cnt  Score   Error  Units
HashUtilBenchmark.hash_old_negative             avgt    5  4.587 ± 0.084  ns/op
HashUtilBenchmark.hash_old_positive             avgt    5  3.746 ± 0.032  ns/op
HashUtilBenchmark.hash_peterOptimized_negative  avgt    5  4.584 ± 0.058  ns/op
HashUtilBenchmark.hash_peterOptimized_positive  avgt    5  3.770 ± 0.015  ns/op
HashUtilBenchmark.hash_peter_negative           avgt    5  4.561 ± 0.052  ns/op
HashUtilBenchmark.hash_peter_positive           avgt    5  3.771 ± 0.046  ns/op
HashUtilBenchmark.hash_rdinctaskin_negative     avgt    5  4.557 ± 0.039  ns/op
HashUtilBenchmark.hash_rdinctaskin_positive     avgt    5  3.793 ± 0.153  ns/op
```

**Comment by mmedenjak** [comment]
The agreement was received. Merging. Thank you for your contribution @erdinctaskin !

**Pull request merged**